internal/signal/signal_test.go

```diff
@@ -0,0 +1,81 @@
+package signal_test
+
+import (
+	"os"
+	"syscall"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/cyberark/secretless-broker/internal/signal"
+)
+
+func TestSignal(t *testing.T) {
+	t.Run("Signals listeners are notified", func(t *testing.T) {
+		testSignals := []os.Signal{
+			syscall.SIGUSR1,
+			syscall.SIGUSR2,
+		}
+		mockListener := signal.NewExitListenerWithOptions(testSignals...)
+
+		notifications := []string{}
+		handler1 := func() {
+			notifications = append(notifications, "handler1")
+		}
+		handler2 := func() {
+			notifications = append(notifications, "handler2")
+		}
+		handler3 := func() {
+			notifications = append(notifications, "handler3")
+		}
+
+		mockListener.AddHandler(handler1)
+		mockListener.AddHandler(handler2)
+		mockListener.AddHandler(handler3)
+
+		// Sanity check
+		assert.Equal(t, 0, len(notifications))
+
+		go func() {
+			syscall.Kill(syscall.Getpid(), syscall.SIGUSR2)
+		}()
+
+		mockListener.Wait()
+
+		assert.Equal(t, []string{"handler1", "handler2", "handler3"}, notifications)
+	})
+
+	t.Run("Signals listeners are only notified on expected signals", func(t *testing.T) {
+		testSignals := []os.Signal{
+			syscall.SIGUSR1,
+			syscall.SIGUSR1,
+		}
+		mockListener := signal.NewExitListenerWithOptions(testSignals...)
+
+		notifications := []string{}
+		handler1 := func() {
+			notifications = append(notifications, "handler1")
+		}
+		handler2 := func() {
+			notifications = append(notifications, "handler2")
+		}
+
+		mockListener.AddHandler(handler1)
+		mockListener.AddHandler(handler2)
+
+		// Sanity check
+		assert.Equal(t, 0, len(notifications))
+
+		go func() {
+			syscall.Kill(syscall.Getpid(), syscall.SIGUSR2)
+		}()
+
+		go func() {
+			assert.Equal(t, 0, len(notifications))
+			syscall.Kill(syscall.Getpid(), syscall.SIGUSR1)
+		}()
+
+		// This should block until we verify that nothing was sent
+		mockListener.Wait()
```

Comment:
Couple issues:

1. Logically, the correctness of the test is relying on the thing we're testing to work correctly.
2. Race condition (I think): It's possible that `Kill` is called in the goroutine before `mockListener.Wait()` in which case it will hang forever. 

We need to actually verify that the signals are sent, independently.  A number of things need to happen afaict. Here's one possible idea:

- Start `Wait()` in a go routine, _before_ starting the other goroutines. 
- Update the actual `Wait()` method and add another method to `ExitListener`: `IsWaiting()` (easy change, but you'll need to use another channel for this I believe)
- Within the test, add a channel to receive notifications from the two goroutines above, ie, and add to this channel  after lines 70 and 75.
- In the goroutine where you started `Wait()` you'll have another channel like `waitHasFinished` which gets notified when wait is done executing.
- At the very bottom of the test method, you `<-waitHasFinished`.

This is probably a case of "write the code is clearer than explain the code" but hopefully that makes sense,  if not lmk.

Reply:
I think I can rework this to fix your comment